### ISSUE TITLE
[no-loop-func] New false positive in 4.0.0 when using a global type in a function in a loop

### ISSUE THREAD

**Issue by lencioni** [typescript-eslint/typescript-eslint#2448]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "no-loop-func": "error"
  }
}
```

```TS
let someArray: MyType[] = [];
for (let i = 0; i < 10; i += 1) {
  someArray = someArray.filter((item: MyType) => !!item);
}
```

It is important here that the type is defined globally for the project in a different file. If the type is defined in the same file, the problem does not appear.

**Expected Result**

no-loop-func does not report an error

**Actual Result**

no-loop-func reports the following error:

> Function declared in a loop contains unsafe references to variable(s) 'MyType'.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.7.0` |
| `node`                             | `12.16.2` |

I suspect this could be related to https://github.com/typescript-eslint/typescript-eslint/pull/2039 maybe?

**Comment by bradzacher** [comment]
I didn't even know that rule used scope analysis! I missed it during my analysis of the ESLint codebase.

This looks like a problem because the base rule assumes that global variables are mutable.
We can extend this rule into this project and add support for ignoring type locations
